src/bt-hh-io.c

```diff
@@ -0,0 +1,729 @@
+/*
+ * Copyright (C) 2016  Mozilla Foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <assert.h>
+#include <fdio/task.h>
+#include <hardware/bluetooth.h>
+#include <hardware/bt_hh.h>
+#include <stdlib.h>
+#include "compiler.h"
+#include "log.h"
+#include "bt-proto.h"
+#include "bt-pdubuf.h"
+#include "bt-core-io.h"
+#include "bt-hh-io.h"
+
+enum {
+  /* commands/responses */
+  OPCODE_CONNECT = 0x01,
+  OPCODE_DISCONNECT = 0x02,
+  OPCODE_VIRTUAL_UNPLUG = 0x03,
+  OPCODE_SET_INFO = 0x04,
+  OPCODE_GET_PROTOCOL = 0x05,
+  OPCODE_SET_PROTOCOL = 0x06,
+  OPCODE_GET_REPORT = 0x07,
+  OPCODE_SET_REPORT = 0x08,
+  OPCODE_SEND_DATA = 0x09,
+  /* notifications */
+  OPCODE_CONNECTION_STATE_NTF = 0x81,
+  OPCODE_HID_INFO_NTF = 0x82,
+  OPCODE_PROTOCOL_MODE_NTF = 0x83,
+  OPCODE_IDLE_TIME_NTF = 0x84,
+  OPCODE_GET_REPORT_NTF = 0x85,
+  OPCODE_VIRTUAL_UNPLUG_NTF = 0x86
+#if ANDROID_VERSION >= 21
+  ,
+  OPCODE_HANDSHAKE_NTF = 0x87
+#endif
+};
+
+static void (*send_pdu)(struct pdu_wbuf* wbuf);
+static const bthh_interface_t* bthh_interface;
+
+static enum ioresult
+send_ntf_pdu(void* data)
+{
+  /* send notification on I/O thread */
+  if (!send_pdu) {
+    ALOGE("send_pdu is NULL");
+    return IO_OK;
+  }
+  send_pdu(data);
+  return IO_OK;
+}
+
+/*
+ * Protocol helpers
+ */
+
+static long
+read_bthh_hid_info_t(const struct pdu* pdu, unsigned long offset,
+                     bthh_hid_info_t* hid_info)
+{
+  uint8_t sub_class, app_id, ctry_code;
+  uint16_t attr_mask, vendor_id, product_id, version, dl_len;
+
+  long off = read_pdu_at(pdu, offset, "SCCSSSCSS", &attr_mask, &sub_class,
+                                                   &app_id, &vendor_id,
+                                                   &product_id, &version,
+                                                   &ctry_code, &dl_len);
+  if (off < 0)
+    return -1;
+
+  hid_info->attr_mask = attr_mask;
+  hid_info->sub_class = sub_class;
+  hid_info->app_id = app_id;
+  hid_info->vendor_id = vendor_id;
+  hid_info->product_id = product_id;
+  hid_info->version = version;
+  hid_info->ctry_code = ctry_code;
+  hid_info->dl_len = dl_len;
```

Comment:
This looks better now. We recently added Q_BLUETOOTH to the build script. It's set when we're building against CAF headers. So if Q_BLUETOOTH is defined, you should set priority to some meaningful value here.